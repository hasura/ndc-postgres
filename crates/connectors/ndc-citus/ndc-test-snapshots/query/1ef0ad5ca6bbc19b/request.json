{
  "collection": "citus_tables",
  "query": {
    "fields": {
      "access_method": {
        "type": "column",
        "column": "access_method"
      },
      "citus_table_type": {
        "type": "column",
        "column": "citus_table_type"
      },
      "colocation_id": {
        "type": "column",
        "column": "colocation_id"
      },
      "distribution_column": {
        "type": "column",
        "column": "distribution_column"
      },
      "shard_count": {
        "type": "column",
        "column": "shard_count"
      },
      "table_name": {
        "type": "column",
        "column": "table_name"
      },
      "table_owner": {
        "type": "column",
        "column": "table_owner"
      },
      "table_size": {
        "type": "column",
        "column": "table_size"
      }
    },
    "limit": 10,
    "order_by": {
      "elements": [
        {
          "order_direction": "desc",
          "target": {
            "type": "column",
            "name": "table_size",
            "path": []
          }
        },
        {
          "order_direction": "desc",
          "target": {
            "type": "column",
            "name": "colocation_id",
            "path": []
          }
        }
      ]
    }
  },
  "arguments": {},
  "collection_relationships": {}
}