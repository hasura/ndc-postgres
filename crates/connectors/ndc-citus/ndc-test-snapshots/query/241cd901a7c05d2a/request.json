{
  "collection": "Genre",
  "query": {
    "fields": {
      "GenreId": {
        "type": "column",
        "column": "GenreId"
      },
      "Name": {
        "type": "column",
        "column": "Name"
      }
    },
    "limit": 10,
    "where": {
      "type": "binary_comparison_operator",
      "column": {
        "type": "column",
        "name": "Name",
        "path": []
      },
      "operator": {
        "type": "equal"
      },
      "value": {
        "type": "scalar",
        "value": "Jazz"
      }
    }
  },
  "arguments": {},
  "collection_relationships": {}
}