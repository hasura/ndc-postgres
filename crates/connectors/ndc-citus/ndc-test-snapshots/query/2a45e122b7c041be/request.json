{
  "collection": "PlaylistTrack",
  "query": {
    "fields": {
      "PlaylistId": {
        "type": "column",
        "column": "PlaylistId"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      }
    },
    "limit": 10
  },
  "arguments": {},
  "collection_relationships": {}
}