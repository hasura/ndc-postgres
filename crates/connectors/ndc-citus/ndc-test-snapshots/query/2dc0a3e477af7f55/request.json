{
  "collection": "MediaType",
  "query": {
    "fields": {
      "MediaTypeId": {
        "type": "column",
        "column": "MediaTypeId"
      },
      "Name": {
        "type": "column",
        "column": "Name"
      }
    },
    "limit": 10,
    "order_by": {
      "elements": []
    }
  },
  "arguments": {},
  "collection_relationships": {}
}