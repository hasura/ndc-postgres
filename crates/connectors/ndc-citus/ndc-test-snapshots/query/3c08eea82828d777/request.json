{
  "collection": "PlaylistTrack",
  "query": {
    "fields": {
      "PlaylistId": {
        "type": "column",
        "column": "PlaylistId"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      }
    },
    "limit": 10,
    "where": {
      "type": "binary_comparison_operator",
      "column": {
        "type": "column",
        "name": "PlaylistId",
        "path": []
      },
      "operator": {
        "type": "equal"
      },
      "value": {
        "type": "scalar",
        "value": 1
      }
    }
  },
  "arguments": {},
  "collection_relationships": {}
}