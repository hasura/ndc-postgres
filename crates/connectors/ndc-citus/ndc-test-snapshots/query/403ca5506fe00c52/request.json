{
  "collection": "Track",
  "query": {
    "fields": {
      "AlbumId": {
        "type": "column",
        "column": "AlbumId"
      },
      "Bytes": {
        "type": "column",
        "column": "Bytes"
      },
      "Composer": {
        "type": "column",
        "column": "Composer"
      },
      "GenreId": {
        "type": "column",
        "column": "GenreId"
      },
      "MediaTypeId": {
        "type": "column",
        "column": "MediaTypeId"
      },
      "Milliseconds": {
        "type": "column",
        "column": "Milliseconds"
      },
      "Name": {
        "type": "column",
        "column": "Name"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      },
      "UnitPrice": {
        "type": "column",
        "column": "UnitPrice"
      }
    },
    "limit": 10,
    "where": {
      "type": "binary_comparison_operator",
      "column": {
        "type": "column",
        "name": "GenreId",
        "path": []
      },
      "operator": {
        "type": "equal"
      },
      "value": {
        "type": "scalar",
        "value": 1
      }
    }
  },
  "arguments": {},
  "collection_relationships": {}
}