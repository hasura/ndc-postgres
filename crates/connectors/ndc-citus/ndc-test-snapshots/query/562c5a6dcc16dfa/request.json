{
  "collection": "PlaylistTrack",
  "query": {
    "fields": {
      "PlaylistId": {
        "type": "column",
        "column": "PlaylistId"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      }
    },
    "limit": 10,
    "order_by": {
      "elements": [
        {
          "order_direction": "desc",
          "target": {
            "type": "column",
            "name": "PlaylistId",
            "path": []
          }
        }
      ]
    }
  },
  "arguments": {},
  "collection_relationships": {}
}