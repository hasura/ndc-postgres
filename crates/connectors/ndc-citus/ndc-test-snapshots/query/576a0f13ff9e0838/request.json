{
  "collection": "Track",
  "query": {
    "fields": {
      "AlbumId": {
        "type": "column",
        "column": "AlbumId"
      },
      "Bytes": {
        "type": "column",
        "column": "Bytes"
      },
      "Composer": {
        "type": "column",
        "column": "Composer"
      },
      "GenreId": {
        "type": "column",
        "column": "GenreId"
      },
      "MediaTypeId": {
        "type": "column",
        "column": "MediaTypeId"
      },
      "Milliseconds": {
        "type": "column",
        "column": "Milliseconds"
      },
      "Name": {
        "type": "column",
        "column": "Name"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      },
      "UnitPrice": {
        "type": "column",
        "column": "UnitPrice"
      }
    },
    "limit": 10,
    "order_by": {
      "elements": [
        {
          "order_direction": "desc",
          "target": {
            "type": "column",
            "name": "Composer",
            "path": []
          }
        }
      ]
    }
  },
  "arguments": {},
  "collection_relationships": {}
}