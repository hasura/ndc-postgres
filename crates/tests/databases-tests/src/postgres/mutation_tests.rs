#[cfg(test)]
/// create a fresh db then run a query against it
mod basic {
    use super::super::common;
    use tests_common::ndc_metadata::{clean_up_ndc_metadata, create_fresh_ndc_metadata};
    use tests_common::request::{run_mutation, run_query};

    #[tokio::test]
    async fn delete_playlist() {
        let ndc_metadata = create_fresh_ndc_metadata(
            common::CONNECTION_STRING,
            common::CHINOOK_NDC_METADATA_PATH_V3,
        )
        .await
        .unwrap();

        let result = run_mutation(
            tests_common::router::create_router_from_ndc_metadata(&ndc_metadata.ndc_metadata_path)
                .await,
            "delete_playlist",
        )
        .await;

        clean_up_ndc_metadata(ndc_metadata).await.unwrap();
        insta::assert_json_snapshot!(result)
    }

    #[tokio::test]
    async fn insert_artist_album() {
        let ndc_metadata = create_fresh_ndc_metadata(
            common::CONNECTION_STRING,
            common::CHINOOK_NDC_METADATA_PATH_V3,
        )
        .await
        .unwrap();

        let router =
            tests_common::router::create_router_from_ndc_metadata(&ndc_metadata.ndc_metadata_path)
                .await;

        let mutation_result = run_mutation(router.clone(), "insert_artist_album").await;

        // expect the artist to be returned, because we inserted it successfully.
        let selection_result = run_query(router, "mutations/select_specific_artist").await;

        let result = (mutation_result, selection_result);

        clean_up_ndc_metadata(ndc_metadata).await.unwrap();
        insta::assert_json_snapshot!(result)
    }

    #[tokio::test]
    async fn delete_invoice_line() {
        let ndc_metadata = create_fresh_ndc_metadata(
            common::CONNECTION_STRING,
            common::CHINOOK_NDC_METADATA_PATH_V3,
        )
        .await
        .unwrap();

        let result = run_mutation(
            tests_common::router::create_router_from_ndc_metadata(&ndc_metadata.ndc_metadata_path)
                .await,
            "delete_invoice_line",
        )
        .await;

        clean_up_ndc_metadata(ndc_metadata).await.unwrap();
        insta::assert_json_snapshot!(result)
    }

    #[tokio::test]
    async fn v1_insert_custom_dog() {
        let ndc_metadata = create_fresh_ndc_metadata(
            common::CONNECTION_STRING,
            common::CHINOOK_NDC_METADATA_PATH_V3,
        )
        .await
        .unwrap();

        let router =
            tests_common::router::create_router_from_ndc_metadata(&ndc_metadata.ndc_metadata_path)
                .await;

        let mutation_result = run_mutation(router.clone(), "v1_insert_custom_dog").await;

        let result = mutation_result;

        clean_up_ndc_metadata(ndc_metadata).await.unwrap();
        insta::assert_json_snapshot!(result)
    }
}

#[cfg(test)]
mod negative {
    use super::super::common;
    use tests_common::ndc_metadata::{clean_up_ndc_metadata, create_fresh_ndc_metadata};
    use tests_common::request::{run_mutation_fail, run_query, StatusCode};

    #[tokio::test]
    /// Check that the second statement fails on duplicate key constraint,
    /// and that it rolls back the first statement.
    async fn insert_artist_album_bad() {
        let ndc_metadata = create_fresh_ndc_metadata(
            common::CONNECTION_STRING,
            common::CHINOOK_NDC_METADATA_PATH_V3,
        )
        .await
        .unwrap();

        let router =
            tests_common::router::create_router_from_ndc_metadata(&ndc_metadata.ndc_metadata_path)
                .await;

        let mutation_result = run_mutation_fail(
            router.clone(),
            "insert_artist_album_bad",
            StatusCode::FORBIDDEN,
        )
        .await;

        // expect no rows returned because first operation was rolled back.
        let selection_result = run_query(router, "mutations/select_specific_artist").await;

        let result = (mutation_result, selection_result);

        clean_up_ndc_metadata(ndc_metadata).await.unwrap();
        insta::assert_json_snapshot!(result);
    }

    #[tokio::test]
    /// Check that insert fails due to missing column.
    async fn v1_insert_custom_dog_missing_column() {
        let ndc_metadata = create_fresh_ndc_metadata(
            common::CONNECTION_STRING,
            common::CHINOOK_NDC_METADATA_PATH_V3,
        )
        .await
        .unwrap();

        let router =
            tests_common::router::create_router_from_ndc_metadata(&ndc_metadata.ndc_metadata_path)
                .await;

        let mutation_result = run_mutation_fail(
            router.clone(),
            "v1_insert_custom_dog_missing_column",
            StatusCode::BAD_REQUEST,
        )
        .await;

        let result = mutation_result;

        clean_up_ndc_metadata(ndc_metadata).await.unwrap();
        insta::assert_json_snapshot!(result);
    }
}
