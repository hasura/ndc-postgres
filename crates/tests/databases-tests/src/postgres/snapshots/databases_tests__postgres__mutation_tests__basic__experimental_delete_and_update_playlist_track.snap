---
source: crates/tests/databases-tests/src/postgres/mutation_tests.rs
expression: result
---
{
  "operation_results": [
    {
      "type": "procedure",
      "result": {
        "returning": [
          {
            "playlist_id": 8,
            "track_id": 90
          }
        ],
        "affected_rows": 1
      }
    },
    {
      "type": "procedure",
      "result": {
        "returning": [
          {
            "track_id": 89,
            "playlist_id": 7
          }
        ],
        "affected_rows": 1
      }
    }
  ]
}
